llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
   return Res;
 }
 
-static string_vector
-saveSpecConstantIDMaps(const std::vector<SpecIDMapTy> &Maps) {
-  string_vector Res;
+// Each fallback device library corresponds to one bit in "require mask" which
+// is an unsigned int32. getDeviceLibBit checks which fallback device library
+// is required for FuncName and returns the corresponding bit. The corresponding
+// mask for each fallback device library is:
+// fallback-cassert:      0x1
+// fallback-cmath:        0x2
+// fallback-cmath-fp64:   0x4
+// fallback-complex:      0x8
+// fallback-complex-fp64: 0x10
+static uint32_t getDeviceLibBits(const std::string &FuncName) {
+
+  if (DeviceLibFuncMap.count(FuncName) == 0)
+    return 0;
+  else
+    return 0x1 << (DeviceLibFuncMap[FuncName] - cl_intel_devicelib_assert);
+}
+
+// For each device image module, we go through all functions which meets
+// 1. The function name has prefix "__devicelib_"
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const Module &M) {
+  // 0x1 means sycl runtime will link and load libsycl-fallback-assert.spv as
+  // default. In fact, default link assert spv is not necessary but dramatic
+  // perf regression is observed if we don't link any device library. The perf
+  // regression is caused by a clang issue.
+  uint32_t ReqMask = 0x1;
+  uint32_t DeviceLibBits = 0;
+  for (const Function &SF : M) {
+    if (SF.getName().startswith(DEVICELIB_FUNC_PREFIX) &&
+        (SF.getCallingConv() == CallingConv::SPIR_FUNC) && SF.isDeclaration()) {
```

Comment:
`SF.getCallingConv() == CallingConv::SPIR_FUNC`
this should really be an assert as we don't really expect non-spirv functions with matching prefix

Reply:
Hi, @kbobrovs 
I agree that we don't expect any non-spirv functions with matching prefix, need we add a note in document telling users that __devicelib_* names are used by DPC++ compiler and they can't use those names in their code?
Thank you very much.

Reply:
Users can't define names starting with "__" anyway - those are reserved for system/compiler, so I don't see a problem. Adding to the doc won't hurt, though :)

Reply:
OK, got it.
Already updated the patch to use assert for (SF.getCallingConv() == CallingConv::SPIR_FUNC).